File: src/main/java/chiloven/lukosbot2/commands/wikis/McWikiCommand.java
package chiloven.lukosbot2.commands.wikis;

import chiloven.lukosbot2.core.CommandSource;
import chiloven.lukosbot2.model.Attachment;
import chiloven.lukosbot2.model.MessageOut;
import chiloven.lukosbot2.util.feature.WebScreenshot;
import chiloven.lukosbot2.util.feature.WebToMarkdown;
import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.arguments.StringArgumentType;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.mojang.brigadier.builder.RequiredArgumentBuilder;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jsoup.Connection;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.springframework.stereotype.Service;

import static com.mojang.brigadier.arguments.StringArgumentType.greedyString;

/**
 * The /mcwiki command for Minecraft Wiki title+lead, markdown conversion, and screenshots.
 *
 * @author Chiloven945
 */
@Service
public class McWikiCommand implements WikiishCommand {
    private static final Logger log = LogManager.getLogger(McWikiCommand.class);

    /**
     * Fetch title and lead paragraph(s) from a Minecraft Wiki article.
     *
     * @param url The full URL of the Minecraft Wiki article.
     * @return {@link TitleLead} object containing the title and lead text.
     * @throws Exception if fetching or parsing fails.
     */
    private static TitleLead fetchTitleAndLead(String url) throws Exception {
        Connection conn = Jsoup.connect(url)
                .userAgent("Mozilla/5.0 (compatible; LukosBot/1.0)")
                .timeout(15000);

        Document doc = conn.get();
        String title = textOrEmpty(doc.selectFirst("h1#firstHeading"));

        // MediaWiki 正文容器通常位于 #mw-content-text .mw-parser-output
        Element container = doc.selectFirst("#mw-content-text .mw-parser-output");
        if (container == null) {
            // 后备
            container = doc.selectFirst("#content");
        }
        if (container == null) {
            return new TitleLead(title, "");
        }

        StringBuilder sb = new StringBuilder();
        int added = 0;
        for (Element p : container.select("> p")) {
            String t = p.text().trim();
            if (t.isEmpty()) continue;
            if (!sb.isEmpty()) sb.append("\n\n");
            sb.append(t);
            added++;
            if (added >= 2 || sb.length() >= 500) break; // 控制简介长度
        }
        return new TitleLead(title, sb.toString().trim());
    }

    /**
     * Get trimmed text of an element, or empty string if null.
     *
     * @param e The Jsoup Element.
     * @return Trimmed text or empty string.
     */
    private static String textOrEmpty(Element e) {
        return e == null ? "" : e.text().trim();
    }

    @Override
    public String defaultLang() {
        return "zh";
    }

    @Override
    public String pathPrefix() {
        return "/w/";
    }

    @Override
    public String domainRoot() {
        return "minecraft.wiki";
    }

    @Override
    public String name() {
        return "mcwiki";
    }

    @Override
    public String description() {
        return "Minecraft Wiki 工具：返回标题+简介，或整页 Markdown";
    }

    @Override
    public String usage() {
        return """
                用法：
                `/mcwiki <minecraftwiki_link|article>`    # 返回“标题\\n简介”
                `/mcwiki md <minecraftwiki_link|article>` # 抓取整页并转为 Markdown 文件
                示例：
                /mcwiki 僵尸猪灵
                /mcwiki md https://zh.minecraft.wiki/w/僵尸猪灵
                /mcwiki en:Zombie_Piglin
                """;
    }

    @Override
    public void register(CommandDispatcher<CommandSource> dispatcher) {
        dispatcher.register(
                LiteralArgumentBuilder.<CommandSource>literal(name())
                        // /mcwiki <link>
                        .then(RequiredArgumentBuilder.<CommandSource, String>argument("link", greedyString())
                                .executes(ctx -> {
                                    String link = StringArgumentType.getString(ctx, "link").trim();
                                    runSummary(ctx.getSource(), link);
                                    return 1;
                                })
                        )
                        // /mcwiki md <link>
                        .then(LiteralArgumentBuilder.<CommandSource>literal("md")
                                .then(RequiredArgumentBuilder.<CommandSource, String>argument("link", greedyString())
                                        .executes(ctx -> {
                                            String link = StringArgumentType.getString(ctx, "link").trim();
                                            runMarkdown(ctx.getSource(), link);
                                            return 1;
                                        })
                                )
                        )
                        // /mcwiki ss <link>
                        .then(LiteralArgumentBuilder.<CommandSource>literal("ss")
                                .then(RequiredArgumentBuilder.<CommandSource, String>argument("link", greedyString())
                                        .executes(ctx -> {
                                            String link = StringArgumentType.getString(ctx, "link").trim();
                                            runScreenshot(ctx.getSource(), link);
                                            return 1;
                                        })
                                )
                        )
                        // 无参时给出用法
                        .executes(ctx -> {
                            ctx.getSource().reply(usage());
                            return 1;
                        })
        );
    }

    /**
     * Run the summary subcommand: fetch title and lead paragraph(s).
     *
     * @param src         The command source.
     * @param linkOrTitle The Minecraft Wiki link or article title.
     */
    private void runSummary(CommandSource src, String linkOrTitle) {
        try {
            String url = normalize(linkOrTitle);
            if (isNot(url)) {
                src.reply("仅支持 Minecraft Wiki 链接或条目名。示例：/mcwiki https://zh.minecraft.wiki/w/下界合金");
                return;
            }

            var summary = fetchTitleAndLead(url);
            if (summary.title().isBlank()) {
                src.reply("未抓到有效内容。请检查条目是否存在。");
                return;
            }

            String text = summary.title() + "\n" + summary.lead();
            MessageOut out = MessageOut.text(src.in().addr(), text);
            src.reply(out);
        } catch (Exception e) {
            log.warn("McWiki summary failed: {}", linkOrTitle, e);
            src.reply("获取失败：" + e.getMessage());
        }
    }

    /**
     * Run the markdown subcommand: fetch and convert to Markdown.
     *
     * @param src         The command source.
     * @param linkOrTitle The Minecraft Wiki link or article title.
     */
    private void runMarkdown(CommandSource src, String linkOrTitle) {
        try {
            String url = normalize(linkOrTitle);
            if (isNot(url)) {
                src.reply("仅支持 Minecraft Wiki 链接或条目名。示例：/mcwiki md https://zh.minecraft.wiki/w/下界合金");
                return;
            }

            var md = WebToMarkdown.fetchAndConvertWithSelectors(
                    url,
                    "h1#firstHeading",                // 标题选择器（与 MediaWiki 一致）
                    "#content, #mw-content-text",     // 内容区主容器（两个择一）
                    "mcwiki"                          // 默认文件名前缀
            );

            MessageOut out = MessageOut.text(src.in().addr(), "已转换为 Markdown")
                    .with(Attachment.fileBytes(md.filename(), md.bytes(), md.mime()));
            src.reply(out);
        } catch (Exception e) {
            log.warn("McWiki md failed: {}", linkOrTitle, e);
            src.reply("转换失败：" + e.getMessage());
        }
    }

    /**
     * Run the screenshot subcommand: capture a screenshot of the article.
     *
     * @param src         The command source.
     * @param linkOrTitle The Minecraft Wiki link or article title.
     */
    private void runScreenshot(CommandSource src, String linkOrTitle) {
        try {
            String url = normalize(linkOrTitle);
            if (isNot(url)) {
                src.reply("仅支持 Minecraft Wiki 链接或条目名。示例：/mcwiki ss https://zh.minecraft.wiki/w/下界合金");
                return;
            }

            var cd = WebScreenshot.screenshotMcWiki(url);

            MessageOut out = MessageOut.text(src.in().addr(), "截图如下：")
                    .with(Attachment.imageBytes(cd.filename(), cd.bytes(), cd.mime()));
            src.reply(out);

        } catch (Exception e) {
            log.warn("McWiki screenshot failed: {}", linkOrTitle, e);
            src.reply("截图失败：" + e.getMessage());
        }
    }

    /**
     * Title and lead paragraph(s) record.
     *
     * @param title the article title
     * @param lead  the lead paragraph(s)
     */
    private record TitleLead(String title, String lead) {
    }
}

File: src/main/java/top/chiloven/lukosbot2/commands/bot/wikis/McWikiCommand.kt
package top.chiloven.lukosbot2.commands.bot.wikis

import org.apache.logging.log4j.LogManager
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.stereotype.Service
import top.chiloven.lukosbot2.Constants
import top.chiloven.lukosbot2.core.command.bot.CommandSource
import top.chiloven.lukosbot2.core.command.definition.dsl.arg
import top.chiloven.lukosbot2.core.command.definition.dsl.botCommand
import top.chiloven.lukosbot2.core.command.definition.dsl.oneOf
import top.chiloven.lukosbot2.core.model.message.media.BytesRef
import top.chiloven.lukosbot2.core.model.message.outbound.OutFile
import top.chiloven.lukosbot2.core.model.message.outbound.OutImage
import top.chiloven.lukosbot2.core.model.message.outbound.OutboundMessage
import top.chiloven.lukosbot2.util.feature.WebScreenshot
import top.chiloven.lukosbot2.util.feature.WebToMarkdown
import java.time.Duration

@Service
@ConditionalOnProperty(
    prefix = "lukos.commands.control",
    name = ["mcwiki"],
    havingValue = "true",
    matchIfMissing = true
)
class McWikiCommand : IWikiishCommand {

    private val log = LogManager.getLogger(McWikiCommand::class.java)

    override fun defaultLang() = "zh"
    override fun pathPrefix() = "/w/"
    override fun domainRoot() = "minecraft.wiki"

    private val commandDefinition = botCommand("mcwiki") {
        description = "查询 Minecraft Wiki，支持简介、截图和导出 Markdown"

        literal("md") {
            description = "抓取整页并转为 Markdown 文件"
            raw("link") { link -> runMarkdown(source, link.trim()) }
            param("link", "Minecraft Wiki 链接")
            param("article", "词条名（可用 `en:` 前缀指定语言）")
        }

        literal("ss") {
            description = "生成页面截图"
            raw("link") { link -> runScreenshot(source, link.trim()) }
            param("link", "Minecraft Wiki 链接")
            param("article", "词条名（可用 `en:` 前缀指定语言）")
        }

        raw("link", required = false) { link ->
            if (link.isBlank()) sendUsage(source)
            else runSummary(source, link.trim())
        }

        syntax("返回标题和简介", oneOf(arg("link"), arg("article")))
        example(
            "mcwiki 僵尸猪灵",
            "mcwiki en:Zombie_Piglin",
            "mcwiki md https://zh.minecraft.wiki/w/僵尸猪灵",
            "mcwiki ss 下界合金"
        )
        note("词条名支持 `en:` 前缀指定语言，例如：`en:Zombie_Piglin`。")
    }

    override fun definition() = commandDefinition

    private fun runSummary(src: CommandSource, linkOrTitle: String) {
        try {
            val url = normalize(linkOrTitle)
            if (isNot(url)) {
                src.reply("仅支持 Minecraft Wiki 链接或条目名。"); return
            }

            val summary = fetchTitleAndLead(url)
            if (summary.title.isBlank()) {
                src.reply("未获取到有效内容，请检查条目是否存在。"); return
            }
            src.reply(buildString {
                append(summary.title)
                if (summary.lead.isNotBlank()) append('\n').append(summary.lead)
            })
        } catch (e: Exception) {
            log.warn("McWiki summary failed: {}", linkOrTitle, e)
            src.reply("获取失败：${e.message}")
        }
    }

    private fun runMarkdown(src: CommandSource, linkOrTitle: String) {
        try {
            val url = normalize(linkOrTitle)
            if (isNot(url)) {
                src.reply("仅支持 Minecraft Wiki 链接或条目名。"); return
            }

            val md = WebToMarkdown.fetchAndConvertWithSelectors(
                url,
                "h1#firstHeading",
                "#content, #mw-content-text",
                "mcwiki"
            )

            val ref = BytesRef(md.filename(), md.bytes(), md.mime())
            src.reply(OutboundMessage(src.addr(), listOf(OutFile(ref, "已转换为 Markdown。", md.filename(), md.mime()))))
        } catch (e: Exception) {
            log.warn("McWiki md failed: {}", linkOrTitle, e)
            src.reply("转换失败：${e.message}")
        }
    }

    private fun runScreenshot(src: CommandSource, linkOrTitle: String) {
        try {
            val url = normalize(linkOrTitle)
            if (isNot(url)) {
                src.reply("仅支持 Minecraft Wiki 链接或条目名。"); return
            }

            val cd = WebScreenshot.screenshotMcWiki(url)
            val ref = BytesRef(cd.filename(), cd.bytes(), cd.mime())
            src.reply(OutboundMessage(src.addr(), listOf(OutImage(ref, "截图如下。", cd.filename(), cd.mime()))))
        } catch (e: Exception) {
            log.warn("McWiki screenshot failed: {}", linkOrTitle, e)
            src.reply("截图失败：${e.message}")
        }
    }

    private data class TitleLead(val title: String, val lead: String)

    private fun fetchTitleAndLead(url: String): TitleLead {
        val doc = Jsoup.connect(url)
            .userAgent("Mozilla/5.0 (compatible; ${Constants.UA})")
            .timeout(Duration.ofSeconds(15).toMillis().toInt())
            .get()

        val title = doc.selectFirst("h1#firstHeading").textOrEmpty()
        val container =
            doc.selectFirst("#mw-content-text .mw-parser-output") ?: doc.selectFirst("#content") ?: return TitleLead(
                title,
                ""
            )
        val lead = buildString {
            var added = 0
            for (p in container.select("> p")) {
                val t = p.text().trim()
                if (t.isEmpty()) continue
                if (isNotEmpty()) append("\n\n")
                append(t)
                added++
                if (added >= 2 || length >= 500) break
            }
        }.trim()

        return TitleLead(title, lead)
    }

    private fun Element?.textOrEmpty() =
        this?.text()?.trim().orEmpty()

}
